Assets/Xbox Live/Scripts/SignInManager.cs

```diff
@@ -0,0 +1,605 @@
+                    SignInCallbacks = new List<UnityAction<XboxLiveUser, XboxLiveAuthStatus, string>>(),
+                    SignOutCallbacks = new List<UnityAction<XboxLiveUser, XboxLiveAuthStatus, string>>()
+                };
+                this.CurrentPlayers.Add(playerNumber, playerInfo);
+            }
+            return true;
+        }
+        return false;
+    }
+
+    /// <summary>
+    /// Signs in a player. Attempts sign in silently first then sign in.
+    /// </summary>
+    /// <param name="playerNumber">The player number</param>
+    /// <param name="playerInfo">The player info of the the player being signed in</param>
+    /// <returns>A Coroutine</returns>
+    private IEnumerator SignInAsync(int playerNumber, XboxLivePlayerInfo playerInfo)
+    {
+        SignInStatus signInStatus = SignInStatus.Success;
+
+        TaskYieldInstruction<SignInResult> signInSilentlyTask = playerInfo.XboxLiveUser.SignInSilentlyAsync().AsCoroutine();
+        yield return signInSilentlyTask;
+
+        try
+        {
+            signInStatus = signInSilentlyTask.Result.Status;
+        }
+        catch (Exception ex)
+        {
+            if (XboxLiveServicesSettings.Instance.DebugLogsOn)
+            {
+                Debug.LogError("Exception occured: " + ex.Message);
+            }
+
+        }
+
+        if (signInStatus != SignInStatus.Success)
+        {
+            TaskYieldInstruction<SignInResult> signInTask = playerInfo.XboxLiveUser.SignInAsync().AsCoroutine();
+            yield return signInTask;
+
+            try
+            {
+                signInStatus = signInTask.Result.Status;
+            }
+            catch (Exception ex)
+            {
+                if (XboxLiveServicesSettings.Instance.DebugLogsOn)
+                {
+                    Debug.LogError("Exception occured: " + ex.Message);
+                }
+            }
+        }
+
+        try
+        {
+            // Throw any exceptions if needed.
+            if (signInStatus == SignInStatus.Success)
+            {
+                CurrentNumberOfPlayers++;
+                XboxLive.Instance.StatsManager.AddLocalUser(playerInfo.XboxLiveUser);
+                XboxLive.Instance.SocialManager.AddLocalUser(playerInfo.XboxLiveUser, SocialManagerExtraDetailLevel.PreferredColorLevel);
+                this.NotifyAllCallbacks(playerNumber, playerInfo.XboxLiveUser, XboxLiveAuthStatus.Succeeded, null, true);
+            }
+            else
+            {
+                NotifyAllCallbacks(playerNumber, null, XboxLiveAuthStatus.Failed, "Sign In Failed: Player " + playerNumber + " failed. Sign In Status: " + signInStatus);
+            }
+        }
+        catch (Exception ex)
+        {
+            if (XboxLiveServicesSettings.Instance.DebugLogsOn)
+            {
+                Debug.LogError("Exception occured: " + ex.Message);
+            }
+        }
+    }
+
+    private void Update()
+    {
+        if (this.PlayersPendingSignIn.Count > 0)
```

Comment:
This might be better in a coroutine with a timeout. In a scenario where, for some reason the player can't sign in (maybe offline), I would hate to constantly trigger this check and logic over and over.

Reply:
This won't actually wait on the player's input. It just starts the coroutines in the case of MUA. The reason for this is we can't use Coroutines within a task. So in the case of MUA, where we use the Userpicker, we need to basically queue the players that indicated they want to sign in and have update call the coroutines for them. We don't wait for the response though. You'll notice that the actual sign in call done within the if is a coroutine itself.